### ISSUE TITLE
Cache incorrectly deals with request "cache-control: nocache"

### ISSUE THREAD

**Issue by wtarreau** [haproxy/haproxy#2048]
### Detailed Description of the Problem

I got trapped a few times on haproxy.org after an update, forcing a reload to take changes into account, to learn that the contents were not up to date for others. And indeed, that's visible using curl. The problem is that while the cache properly avoids retrieving a response from its storage when the request contains `cache-control: no-cache`, due to using a single flag (`TX_CACHE_IGNORE`) for this, it also prevents the response from updating the cache. Example:

- First request:
```
$ curl -i 127.0.0.1:8001/?s=100
HTTP/1.1 200 
content-length: 100
x-req: size=84, time=0 ms
x-rsp: id=dummy, code=200, cache=1, size=100, time=0 ms (0 real)
cache-control: max-age=1000

.123456789.123456789.123456789.123456789.12345678
.123456789.123456789.123456789.123456789.12345678
```

- second request 4 seconds later, the object comesfrom the cache and is 4-s old:
```
$ curl -i 127.0.0.1:8001/?s=100
HTTP/1.1 200 
content-length: 100
x-req: size=84, time=0 ms
x-rsp: id=dummy, code=200, cache=1, size=100, time=0 ms (0 real)
cache-control: max-age=1000
age: 4
cache-control: max-age=1000

.123456789.123456789.123456789.123456789.12345678
.123456789.123456789.123456789.123456789.12345678
```

- New request 10s later with `cache-control: no-cache`, we indeed get a fresh response:
```
$ curl -i -H "cache-control: no-cache" 127.0.0.1:8001/?s=100
HTTP/1.1 200 
content-length: 100
x-req: size=109, time=0 ms
x-rsp: id=dummy, code=200, cache=1, size=100, time=0 ms (0 real)
cache-control: max-age=1000

.123456789.123456789.123456789.123456789.12345678
.123456789.123456789.123456789.123456789.12345678
```

- Finally new request again retrieving the old copy instead of the updated one:
```
$ curl -i 127.0.0.1:8001/?s=100
HTTP/1.1 200 
content-length: 100
x-req: size=84, time=0 ms
x-rsp: id=dummy, code=200, cache=1, size=100, time=0 ms (0 real)
cache-control: max-age=1000
age: 24
cache-control: max-age=1000

.123456789.123456789.123456789.123456789.12345678
.123456789.123456789.123456789.123456789.12345678
```

### Expected Behavior

The cached contents should be correctly updated with the new response (unless of course the response headers contains headers preventing this).

As a work-around, in order to force updates I'm sending POST requests via curl for the pages, but that's dirty and inconvenient.

### Steps to Reproduce the Behavior

See example above.

### Do you have any idea what may have caused this?

The reason is that we're using the same flag to check the request and the response (`TX_CACHE_IGNORE`). I verified that the issue exists since 2.4 (2.2 was OK).

### Do you have an idea how to solve the issue?

We should instead have one per direction ,e.g. `TX_CACHE_NOLOOKUP`, `TX_CACHE_NOSTORE`.

### What is your configuration?

```haproxy
global
        stats socket /tmp/sock1 level admin mode 666
        stats timeout 1d

defaults
        mode http
        timeout client 10s
        timeout server 10s
        timeout connect 10s

frontend f1
        bind :8181
        stats uri /

listen l1
        bind :8001
        http-request cache-use c1
        http-response add-header cache-control max-age=1000
        http-response cache-store c1
        server s1 127.0.0.1:8000

cache c1
        total-max-size 2048
        max-object-size 1048576000
        max-age 100
```

### Output of `haproxy -vv`

```plain
$ ./haproxy -vv
HAProxy version 2.8-dev4-156a89-33 2023/02/20 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 5.10.70 #1 SMP Sun Oct 3 00:04:32 CEST 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = /f/tc/distcc/remote/x86_64-gcc95_glibc228-linux-gnu-gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_ZLIB= USE_SLZ=1 USE_DEVICEATLAS=1 USE_51DEGREES=1 USE_WURFL=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE=1
  DEBUG   = -DDEBUG_DONT_SHARE_POOLS -DDEBUG_MEMORY_POOLS -DDEBUG_EXPR -DDEBUG_STRICT=2 -DDEBUG_DEV -DDEBUG_MEM_STATS

Feature list : +51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H +DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT +PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY +WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 3.0.1+quic 14 Dec 2021
Running on OpenSSL version : OpenSSL 3.0.1+quic 14 Dec 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.5
Built with the Prometheus exporter as a service
Built with DeviceAtlas support (dummy library only).
Built with 51Degrees Trie support (dummy library).
Built with WURFL support (dummy library version 1.11.2.100)
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.45 2021-06-15
Running on PCRE version : 8.45 2021-06-15
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.5.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Additional Information

As a side note (not directly relevant to this), please also check in `http_action_req_cache_use()`, I find it suspicious that we check that the response to the primary key is complete before checking the secondary key, and that we don't seem to check for this second key's completeness:

```C
	/* We must not use an entry that is not complete. */
	if (res && res->complete) {
		struct appctx *appctx;
		entry_block = block_ptr(res);
		shctx_row_inc_hot(shctx_ptr(cache), entry_block);
		shctx_unlock(shctx_ptr(cache));

		/* In case of Vary, we could have multiple entries with the same
		 * primary hash. We need to calculate the secondary hash in order
		 * to find the actual entry we want (if it exists). */
		if (res->secondary_key_signature) {
			if (!http_request_build_secondary_key(s, res->secondary_key_signature)) {
```

I suspect that it might be possible to return incomplete objects based on a vary header or maybe not return an object for which a vary exists as long as the primary one is not complete, but I could be wrong. I would find it more natural to check `res->complete` after we determine whether there is a secondary key or not.

**Comment by rlebreton** [comment]
Thanks for this very detailed issue, here are two patches for the two issues described.

Concerning the first one, instead of splitting the TX_CACHE_IGNORE flags in two as you suggested, I removed its use on the response because it should never have happened in the first place. The IGNORE flag should only have been used when processing the request, and the TX_CACHEABLE one was already there to express the fact that a response can be cached (regardless of whether the client accepted a cache entry or not).
I spent a bit of time on the second one to see if there where some securities elsewhere to prevent the potential bugs you mentioned but I did not find any. There are indeed some potential bugs, nice catch (expecially since it would have been a pain to debug if someone stumbled into it).
[0001-BUG-MINOR-cache-Cache-response-even-if-request-has-n.txt](https://github.com/haproxy/haproxy/files/10795792/0001-BUG-MINOR-cache-Cache-response-even-if-request-has-n.txt)
[0002-BUG-MINOR-cache-Check-cache-entry-is-complete-in-cas.txt](https://github.com/haproxy/haproxy/files/10795793/0002-BUG-MINOR-cache-Check-cache-entry-is-complete-in-cas.txt)
